[34167] Configurable project columns and gantt link

oliverguenther opened the pull request:
https://community.openproject.com/wp/34167

oliverguenther committed:
Make columns configurable in admin settings

oliverguenther committed:
Generate URL params for gantt view and insert projects

openprojectci commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/op...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/1774f3fc45cb8e6cda546cce327b9f017665ce24/frontend/src/app/modules/admin/types/type-form-configuration.component.ts#L34">frontend/src/app/modules/admin/types/type-form-configuration.component.ts#L34</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

openprojectci commented on frontend/src/app/components/wp-table/external-configuration/external-query-configuration.component.ts:
```diff
@@ -1,35 +1,52 @@
+  disabledTabs?:{ [key:string]:string };
   callback:(newQuery:any) => void;
 }
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/1774f3fc45cb8e6cda546cce327b9f017665ce24/frontend/src/app/components/wp-table/external-configuration/external-query-configuration.component.ts#L15"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

ulferts commented on spec/features/projects/projects_index_spec.rb:
```diff
@@ -202,7 +202,16 @@ def expect_projects_in_order(*projects)
+
+    scenario 'CF columns and filters are visible when added to settings' do
+      Setting.enabled_projects_columns << "cf_#{custom_field.id}"
+      Setting.enabled_projects_columns << "cf_#{invisible_custom_field.id}"
```
Doing this breaks test isolation as the settings are carried over to the next test.

ulferts commented on spec/features/projects/projects_index_spec.rb:
```diff
@@ -847,6 +856,8 @@ def expect_projects_in_order(*projects)
     end
 
     scenario 'allows to alter the order in which projects are displayed' do
+      Setting.enabled_projects_columns << "cf_#{integer_custom_field.id}"
```
Same as above.

ulferts reviewed (changes requested):
A couple of minor things:

* It is possible to configure the query that is applied when the user clicks on the "Open as Gantt view" button. In that configuration, I can select to have the "Project" filter set to some value. Regardless of what the user has selected for the filter,  is then overwritten with the projects currently filtered for. Ideally, we would prevent uselessly setting the value. Admittedly, this is an edge case.
* When all projects are filtered out, the button is still available. Clicking on it, will result in an erroneous query. I suggest to disable the button in such a case. 
* Only the projects of the current page serve as input values for the project filter behind the "Open as Gantt view" button. This is due to the `@projects` scope already being paginated.
* The matrix layout chosen for the selection of the project list was taken over from the work packages settings and suffer from the same shortcomings:
![image](https://user-images.githubusercontent.com/617519/91853211-292d3480-ec62-11ea-9501-7acd5b3b9043.png)
  * Displayed as a matrix when there is no need for it as it is basically a list of checkboxes with their labels. A styling like above would have served the same goal and would allow us to get rid of the matrix: 
![image](https://user-images.githubusercontent.com/617519/91853772-f9caf780-ec62-11ea-86c6-e5a31eed2523.png)
  * The header "Selected Columns" makes no sense
  * The labels are not connected with the checkbox.
  * As a personal preference, I would have listed the "Name" column together with the rest and prevented disabling it. That way, it becomes clearer that it will always be there. 
  * The list does not appear to be ordered.

oliverguenther committed:
Ensure cache is cleared for  settings

oliverguenther committed:
Remove matrix in favor of multiselect

oliverguenther committed:
Show name but disable it

oliverguenther committed:
Add title to explain what the button does and disable if no results

oliverguenther commented:
I addressed all your remarks with the exception of the first one:

> It is possible to configure the query that is applied when the user clicks on the "Open as Gantt view" button. In that configuration, I can select to have the "Project" filter set to some value. Regardless of what the user has selected for the filter, is then overwritten with the projects currently filtered for. Ideally, we would prevent uselessly setting the value. Admittedly, this is an edge case.

I agree with the sentiment, but there's no easy way to restrict the set of allowed filters with the way we handle these external queries. As an easy fix I could disable the filter tab altogether but that would prevent other more useful filters.

ulferts reviewed (changes requested):
Thanks for the fixes, @oliverguenther. 

I fear I spotted a new bug in there which is caused by archived projects being selected. Again, those will result in an invalid query. I propose to remove them from the projects scope.

oliverguenther committed:
Only filter for active projects

oliverguenther commented:
> Thanks for the fixes, @oliverguenther. 
> 
> I fear I spotted a new bug in there which is caused by archived projects being selected. Again, those will result in an invalid query. I propose to remove them from the projects scope.

Good catch. I filter for active projects now.

ulferts reviewed (approved):
Good work :+1:
